astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,860 @@
+        The coordinate (position and velocity) of observer.
+    target : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
+        The coordinate (position and velocity) of target.
+    radial_velocity : `~astropy.units.Quantity`, optional
+        The radial velocity of the target with respect to the observer.
+    redshift : float, optional
+        The relativistic redshift of the target with respect to the observer.
+    doppler_rest : `~astropy.units.Quantity`, optional
+        The rest value to use when expressing the spectral value as a velocity.
+    doppler_convention : str, optional
+        The Doppler convention to use when expressing the spectral value as a velocity.
+    """
+
+    def __new__(cls, value, unit=None,
+                observer=None, target=None,
+                radial_velocity=None, redshift=None,
+                **kwargs):
+
+        obj = super().__new__(cls, value, unit=unit, **kwargs)
+
+        # There are two main modes of operation in this class. Either the
+        # observer and target are both defined, in which case the radial
+        # velocity and redshift are automatically computed from these, or
+        # only one of the observer and target are specified, along with a
+        # manually specified radial velocity or redshift. So if a target and
+        # observer are both specified, we can't also accept a radial velocity
+        # or redshift.
+        if target is not None and observer is not None:
+            if radial_velocity is not None or redshift is not None:
+                raise ValueError("Cannot specify radial velocity or redshift if both "
+                                 "target and observer are specified")
+
+        # We only deal with redshifts here and in the redshift property.
+        # Otherwise internally we always deal with velocities.
+        if redshift is not None:
+            if radial_velocity is not None:
+                raise ValueError("Cannot set both a radial velocity and "
+                                 "redshift on spectral coordinate.")
+            radial_velocity = _redshift_to_velocity(redshift)
+
+        # If we're initializing from an existing SpectralCoord, keep any
+        # parameters that aren't being overridden
+        if observer is None:
+            observer = getattr(value, 'observer', None)
+        if target is None:
+            target = getattr(value, 'target', None)
+
+        # As mentioned above, we should only specify the radial velocity
+        # manually if either or both the observer and target are not
+        # specified.
+        if observer is None or target is None:
+            if radial_velocity is None and redshift is None:
+                radial_velocity = getattr(value, 'radial_velocity', None)
+
+        # Validate the observer and target
+        if observer is not None and not isinstance(observer, (SkyCoord, BaseCoordinateFrame)):
```

Comment:
I'd like to punt on this for now (frames without data as observers) - for the velocity frame transformation it's fine since the position doesn't matter, but for this I think I'd rather we discussed it in a separate PR. In the interest of wrapping this up for feature freeze I'd be tempted to do it as it's implemented right now and then potentially allow these extra use cases in future.